skimage/segmentation/random_walker_segmentation.py

```diff
@@ -77,163 +74,195 @@ def _make_graph_edges_3d(n_x, n_y, n_z):
+    lap_sparse = rows[:, unlabeled_indices]
+    B = -rows[:, seeds_indices]
+
+    seeds = labels[seeds_mask]
+    seeds_mask = sparse.csc_matrix(np.hstack(
+        [np.atleast_2d(seeds == lab).T for lab in range(1, nlabels + 1)]))
+    rhs = B.dot(seeds_mask)
+
     return lap_sparse, rhs
 
 
-def _mask_edges_weights(edges, weights, mask):
-    """
-    Remove edges of the graph connected to masked nodes, as well as
-    corresponding weights of the edges.
-    """
-    mask0 = np.hstack((mask[:, :, :-1].ravel(), mask[:, :-1].ravel(),
-                       mask[:-1].ravel()))
-    mask1 = np.hstack((mask[:, :, 1:].ravel(), mask[:, 1:].ravel(),
-                       mask[1:].ravel()))
-    ind_mask = np.logical_and(mask0, mask1)
-    edges, weights = edges[:, ind_mask], weights[ind_mask]
-    max_node_index = edges.max()
-    # Reassign edges labels to 0, 1, ... edges_number - 1
-    order = np.searchsorted(np.unique(edges.ravel()),
-                            np.arange(max_node_index + 1))
-    edges = order[edges.astype(np.int64)]
-    return edges, weights
-
-
-def _build_laplacian(data, spacing, mask=None, beta=50,
-                     multichannel=False):
-    l_x, l_y, l_z = tuple(data.shape[i] for i in range(3))
-    edges = _make_graph_edges_3d(l_x, l_y, l_z)
-    weights = _compute_weights_3d(data, spacing, beta=beta, eps=1.e-10,
-                                  multichannel=multichannel)
-    if mask is not None:
-        edges, weights = _mask_edges_weights(edges, weights, mask)
-    lap = _make_laplacian_sparse(edges, weights)
-    del edges, weights
-    return lap
+def _solve_linear_system(lap_sparse, B, tol, mode):
 
+    if mode is None:
+        mode = 'cg_j'
 
-def _check_isolated_seeds(labels):
-    """
-    Prune isolated seed pixels to prevent labeling errors, and
-    return coordinates and label values of isolated seeds, so
-    that it is possible to put labels back in random walker output.
-    """
-    fill = ndi.binary_propagation(labels == 0, mask=(labels >= 0))
-    isolated = np.logical_and(labels > 0, np.logical_not(fill))
-    inds = np.nonzero(isolated)
-    values = labels[inds]
-    labels[inds] = -1
-    return inds, values
+    if mode == 'cg_mg' and not amg_loaded:
+        warn('"cg_mg" not available, it requires pyamg to be installed. '
+             'The "cg_j" mode will be used instead.',
+             stacklevel=2)
+        mode = 'cg_j'
+
+    if mode == 'bf':
+        X = spsolve(lap_sparse, B.toarray()).T
+    else:
+        maxiter = None
+        if mode == 'cg':
+            if UmfpackContext is None:
+                warn('"cg" mode may be slow because UMFPACK is not availabel. '
+                     'Consider building Scipy with UMFPACK or use a '
+                     'preconditioned version of CG ("cg_j" or "cg_mg" modes).',
+                     stacklevel=2)
+            M = None
+        elif mode == 'cg_j':
+            M = sparse.diags(1.0 / lap_sparse.diagonal())
```

Comment:
Beautifully simple :-). Do you confirm that there shouldn't be any zeros on the diagonal because this happens only for isolated pixels, which are taken care of before?

Reply:
Yes, exactly: the diagonal is minus the sum of the extra-diagonal elements of the matrix for each row. These extra-diagonal elements are all negative/non null, the diagonal elements can then not be null.